### DISCUSSION TITLE
Static type assertions for testing

### DISCUSSION THREAD

**Discussion by whitphx** [python/typing#1030] [Q&A] [answered]
Dear all,

Is there some way to statically test the type inferences?

---

For example, assume that we have a function `foo()`, whose return type is complex as below.
```python
def foo(type_, is_list):
    if type_ == "str":
        if is_list:
            return ["foo", "bar"]
        else:
            return "foo"
    elif type_ == "int":
        if is_list:
            return [42, 43]
        else:
            return 42
    ...
```
(It's ridiculous, but just a toy example.)

As a library author, I write type annotations on it to support the library users as below:
```python
@overload
def foo(type_: Literal["str"], is_list: Literal[True]) -> list[str]:
    ...
    
@overload
def foo(type_: Literal["str"], is_list: Literal[False]) -> str:
    ...
```

So, next I want to write tests to check the type annotations are working as expected like below.
```python
# This is pseudo code
assert return_type_of(foo("str", True)) is list[str]
assert return_type_of(foo("str", False)) is str
```

To do it, for now, I define a following helper class,

```python
from typing import Generic, TypeVar

T = TypeVar('T')


class is_(Generic[T]):
    def __init__(self, arg: T) -> None:
        pass
```
use it in the test code as below,
```python
is_[list[str]](foo("str", True))
is_[str](foo("str", False))
```

and check it with mypy.

My question is, is there any standard/authentic way to do such a thing?

---
FYI, In other words, I want to write tests like the TypeScript DefinitelyTyped, for example [this](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/esrever/esrever-tests.ts), where there are assertions on the static types

**Comment by srittau** [comment]
I don't think there is a standard way to do this. In fact, in typeshed issue python/typeshed#1339 has been open for quite a while. While we have other tests in typeshed, we still miss the ability to test complex corner cases directly. But @sobolevn mentioned [pytest-mypy-plugins](https://github.com/typeddjango/pytest-mypy-plugins) in that thread, maybe this is something to look into?

**Reply by sobolevn** [reply]
Yes, I am using (and maintaining) `pytest-mypy-plugins` for all my projects. Some demos:
- https://github.com/typeddjango/django-stubs/tree/master/tests/typecheck
- https://github.com/dry-python/returns/tree/master/typesafety
- https://github.com/dry-python/classes/tree/master/typesafety

I've also covered this in a tutorial: https://sobolevn.me/2019/08/testing-mypy-types

**Comment by whitphx** [comment]
Thank you so much!
It seems to be exactly what I have to look into.
I'm going to check out `pytest-mypy-plugins`

**Comment by whitphx** [comment]
Thank you, but I found `pytest-mypy-plugins` did not work well with my actual project.

Its features such as parameterization or capability to assert type-check errors look great, but it's too hard for me to bootstrap the typing tests with it.
(And it's bit hard to write the test cases in YAML files, not .py)

Thank you again, and it was nice I could know that there is no standard way to test the static type inferences - then I will start with a my own minimum helper class just for my own case.

---
FYI:
The project is https://github.com/whitphx/streamlit-webrtc, and I'm writing a test case like below:

```python
from typing import Generic, TypeVar, Optional

from streamlit_webrtc import webrtc_streamer, VideoProcessorBase  # My library

T = TypeVar('T')


# My helper class for type assertion
class is_(Generic[T]):
    def __init__(self, arg: T) -> None:
        pass


def test() -> None:
    class VideoProcessor(VideoProcessorBase):
        foo: int = 42

        def recv(self, frame):
            return frame

    ctx = webrtc_streamer(key="foo", video_processor_factory=VideoProcessor)

    # I want to test whether the generic class and its type inference work as expected
    is_[Optional[VideoProcessor]](ctx.video_processor)
    is_[None](ctx.audio_processor)
    if ctx.video_processor:
        is_[int](ctx.video_processor.foo)
```

The `mypy.ini` is
```yaml
[mypy]
python_version = 3.8
ignore_missing_imports = True
```

I wrote a test case with `pytest-mypy-plugins` as below,
```python
- case: test
  main: |
    from streamlit_webrtc import webrtc_streamer, VideoProcessorBase

    class VideoProcessor(VideoProcessorBase):
        foo: int = 42

        def recv(self, frame):
            return frame

    ctx = webrtc_streamer(key="foo", video_processor_factory=VideoProcessor)
    reveal_type(ctx.video_processor)  # N: Revealed type is 'VideoProcessor'

```
but encountered the following massive errors - doesn't it work correctly when the dependencies do not have type definitions?
```shell
E   pytest_mypy_plugins.utils.TypecheckAssertionError: Invalid output:
E   Actual:
E     ../../../../../../path/to/streamlit-webrtc/.venv/lib/python3.8/site-packages/_pytest/_argcomplete:103: error: Cannot find implementation or library stub for module named "argcomplete.completers" (diff)
E     ../../../../../../path/to/streamlit-webrtc/.venv/lib/python3.8/site-packages/_pytest/_argcomplete:103: error: Cannot find implementation or library stub for module named "argcomplete" (diff)
E     ../../../../../../path/to/streamlit-webrtc/.venv/lib/python3.8/site-packages/_pytest/_code/code:33: error: Skipping analyzing "pluggy": module is installed, but missing library stubs or py.typed marker (diff)
E     ../../../../../../path/to/streamlit-webrtc/.venv/lib/python3.8/site-packages/_pytest/_io/terminalwriter:70: error: Cannot find implementation or library stub for module named "colorama" (diff)

(...a lot of lines)

E     ../../../../../../path/to/streamlit-webrtc/streamlit_webrtc/receive:8: error: Skipping analyzing "aiortc.mediastreams": module is installed, but missing library stubs or py.typed marker (diff)
E     ../../../../../../path/to/streamlit-webrtc/streamlit_webrtc/relay:1: error: Skipping analyzing "aiortc.contrib.media": module is installed, but missing library stubs or py.typed marker (diff)
E     ../../../../../../path/to/streamlit-webrtc/streamlit_webrtc/webrtc:16: error: Skipping analyzing "aiortc": module is installed, but missing library stubs or py.typed marker (diff)
E     ../../../../../../path/to/streamlit-webrtc/streamlit_webrtc/webrtc:17: error: Skipping analyzing "aiortc.contrib.media": module is installed, but missing library stubs or py.typed marker (diff)
E     ../../../../../../path/to/streamlit-webrtc/streamlit_webrtc/webrtc:18: error: Skipping analyzing "aiortc.mediastreams": module is installed, but missing library stubs or py.typed marker (diff)
```

**Reply by erictraut** [reply]
Your helper class `_is` is a clever approach, but it has the downside that a type of `Any` will not produce an error even when that's probably not the intent. So that creates a pretty big hole in your testing.

In pyright's internal tests, I use `reveal_type` for performing static type assertions. Pyright's implementation of `reveal_type` returns a literal string that represents the textual form of the type being revealed. This allows me to do the following:

```python
x: Literal["bool"] = reveal_type(True | False)
```

This approach has a couple of problems though, and I've been thinking about the best way to address these.

First, the `reveal_type` call behavior isn't standardized, and mypy implements it such that it simply returns the value passed in. There's a [proposal to standardize it](https://mail.python.org/archives/list/typing-sig@python.org/thread/YIZ3BKBQZVBTZH6GUA2EBOSWN7LAHE6M/) (which I support), but this will force me to rethink how we do our testing.

Second, the text representation of a type differs from one type checker to another, so this wouldn't work in some cases. For example, pyright always uses the PEP 604 representations for unions, whereas mypy uses the older-style `Union[]` notation.

```python
def func(x: int | str):
    # Pyright: "int | str"
    # Mypy: "Union[builtins.int, builtins.str]"
    reveal_type(x)
```

For the first of these two issues, maybe we should consider adding a variant of `reveal_type` that is effectively an assert. It could either be a two-parameter form of `reveal_type` or a function with a different name, such as `assert_type`. Both of these would be no-ops at runtime (i.e. they'd simply return the value of the first argument).

```python
reveal_type(True | False, "bool")
```
or

```python
assert_type(True | False, "bool")
```

Alternatively, reveal_type could accept keyword arguments that specify a text representation of a type or a type itself. If a text representation is provided, a type checker would verify the text representation. If a type is provided, a type checker would verify that the type matches exactly.

```python
reveal_type(3, expected_text="Literal[2]") # Error: type doesn't match
reveal_type([3], expected_type=list[str]) # Error: type doesn't match
```

If we did this, I'd propose that we define `reveal_type` as follows:
```python
@overload
def reveal_type(__value: _T) -> _T: ...
@overload
def reveal_type(__value: _T, *, expected_text: str) -> _T: ...
@overload
def reveal_type(__value: _T, *, expected_type: Any) -> _T: ...
```
Thoughts?

--
Eric Traut
Contributor to Pyright & Pylance
Microsoft

**Reply by JelleZijlstra** [reply]
I'd also be interested in standardizing some variant of what you're proposing. But I'd be wary about adding it to the current reveal_type() proposal because this would be rather more complicated to standardize and implement in type checkers.

For comparison, here's how current type checkers implement this:
- pyright uses `x: Literal["bool"] = reveal_type(True | False)` (which won't work with the proposed standardized behavior of `reveal_type`)
- mypy uses `reveal_type(True | False)  # N: Revealed type is "bool"` (mypy's test framework tests exact error messages)
- pyre also uses `reveal_type()` output, e.g. https://github.com/facebook/pyre-check/blob/main/source/analysis/test/integration/assignTest.ml
- pytype does the same, e.g. https://github.com/google/pytype/blob/4f2f2a0c09f914491434ac2a23c94384eecde3b3/pytype/tests/test_exceptions2.py#L98
- pyanalyze uses `assert_is_value(True | False, TypedValue(bool))` (using pyanalyze's internal representation for types)

So the most common approach is to basically string-match the result of a reveal_type() call. However, that is hard to standardize because type formatting will naturally differ across type checkers.

My preference would be to add `assert_type(True | False, bool)`. Using a separate function makes more sense to me than overloading `reveal_type()` for a distinct use case, and I'd want to avoid using strings because they're harder to make portable across type checkers.

**Reply by whitphx** [reply]
Thank you, it all makes sense.

I missed the big pitfall you mentioned that unexpected `Any` cannot be caught with the `is_` assertion.
I will try `pyright` to use this style of assertion: `x: Literal["bool"] = reveal_type(True | False)`.

The 2-arguments version of `reveal_type` or `assert_type` is exactly what I want. In my case, I want to use the `expected_type` assertion.
Though I'm not a specialist in Python typing, I personally support the proposal.

I don't have the preference on the func name (overloading the current `reveal_type` or not), and JelleZijlstra's opinion also makes sense.
Any of them, `assert_type(True | False, bool)` too, is perfect for me.

It's truly good if it is released publicly whichever as a standardized API or a specific API of `pyright`.

**Reply by erictraut** [reply]
Based on this discussion, I decided to change the mechanism used by pyright. The new release (1.1.211, which I just published) now implements the mechanism proposed above. Its treatment of `reveal_type` now accepts optional `expected_type` and `expected_text` keyword parameters. I've also updated all of pyright's tests cases to use the new mechanism, and I'm pretty happy with it. Let me know if this addresses your use case and whether you have any suggestions.

@JelleZijlstra, I'm open to switching to a different standard function name (like `assert_type`) if that is the consensus of the typing community.

Here's an example:
![Screen Shot 2022-01-18 at 12 21 49 AM](https://user-images.githubusercontent.com/7040122/149898328-2dc567eb-2781-49ea-9c10-04a57a831d9d.png)

**Reply by whitphx** [reply]
Great! I can use it now and it's perfect for my purpose, thank you!

BTW, regarding the func name, I changed my mind after I actually wrote and executed some tests.
Now I think `assert_type` is better.
The reason, in addition to JelleZijlstra's, is it prints out the type information to stdout, which is not needed for testing.
`reveal_type()` is originally designed for printing the types, and I think we should create a different method for a different purpose.

Thank you again!

---
Update:
Now I also have another aspect that `reveal_type()` is considered as a general-purpose method for debugging and its functionality such as printing or assertion should be controlled via runtime options of the type checker.
I don't know which is better - I again don't have any preferences on it. Both work at least in my case.

**Reply by JelleZijlstra** [reply]
It did occur to me that we could perhaps use breakpoint() from https://www.python.org/dev/peps/pep-0553/ instead of reveal_type().

**Reply by thejcannon** [reply]
For those who are chained to `mypy`, I just published a tiny plugin which does roughly this (and handles the `Any` case correctly AFAIK)
https://pypi.org/project/mypy-typing-asserts/

Really hope this makes it into `typing` though so the community can get standard support <3

**Comment by rchen152** [comment]
For what it's worth, pytype has separate `reveal_type` and `assert_type` functions: https://google.github.io/pytype/faq.html#can-i-find-out-what-pytype-thinks-the-type-of-my-expression-is. We added `assert_type` in response to a user request pretty much identical to the one here - users wanted to be able to assert on types in their tests.

**Comment by martindemello** [comment]
incidentally, from a dev rather than a user perspective `assert_type` has been useful in simplifying some of our unit tests, since we could go from

```
x = InferTypes("code block")
assertTypes(x, "pyi fragment")
```

to

```
CheckTypes("code block with some assert_type statements included")
```

**Reply by erictraut** [reply]
@JelleZijlstra, what if we simply added `reveal_type`, `reveal_locals` and `assert_type` to the `typing_extensions` library? That would provide standardization across type checkers without the need for a PEP or for adding them to stdlib (the `typing` module).

I think we could adopt the pytype proposal for `assert_type`. My only concern is that it always interprets a str argument as text rather than a forward-declared type expression. That's why I proposed adding separate named parameters `expected_type` and `expected_text` in my proposal above. But maybe forward-declared type expressions aren't important for this use case and I'm overthinking it.

**Reply by JelleZijlstra** [reply]
> @JelleZijlstra, what if we simply added reveal_type, reveal_locals and assert_type to the typing_extensions library?

That's an option, but so far we have stuck to putting only stdlib backports in typing-extensions. Also, I think it would be valuable to have them documented in https://docs.python.org/3/library/typing.html.

> My only concern is that it always interprets a str argument as text rather than a forward-declared type expression. 

My suggestion would be that the type argument should be interpreted like the first argument to cast(), so a string would be interpreted as a forward reference. That's a little different from how pytype currently does it, though.

----

One other related thing I've been thinking about. `assert_type()` is useful for testing that type checkers interpret your library the way you want them to do. Another related thing you might want to do is assert that certain code fails type checking. Could we add a mechanism for that?

Here's a possible spelling:
```
with typing.expect_error(
    # each type checker checks that its own error code matches the expected value,
    # and ignores arguments meant for other type checkers
    pyright_code="reportGeneralTypeIssue",
    pyanalyze_code="incompatible_argument",
):
     len(3)
```
But that feels like an awkward API.

**Reply by erictraut** [reply]
Yeah, that feels pretty awkward. I'd prefer to keep it simple and stick with the three APIs we've discussed above. A context manager feels like overkill. I haven't heard anyone raise a need for that. I'm also not crazy about hard-coding type checker names into a standardized API.

**Reply by AlexWaygood** [reply]
> Another related thing you might want to do is assert that certain code fails type checking.

There was actually a user request for this on twitter the other day: https://twitter.com/tintvrtkovic/status/1481626149792985093?s=21. But it's not something I've ever felt a need for. I agree it seems less important than `assert_type`.

**Reply by sobolevn** [reply]
> that certain code fails type checking

This is a very common problem for mypy plugins. Plugins might have multiple custom failures. For example, we check all of them in our typesafety test cases: https://github.com/dry-python/classes/blob/master/typesafety/test_typeclass/test_validation/test_body.yml#L21-L69

**Reply by JelleZijlstra** [reply]
It would also be useful for typeshed to be able to test complex signatures. For example, it would be nice if I could commit the test case from python/typeshed#6941 into the repo somehow and ensure that it keeps working with all type checkers.

The type checker-specific error codes are probably overkill though, so let's drop that.

**Reply by AlexWaygood** [reply]
Wait, yes, I stand corrected -- it is something [I have _definitely_ felt a need for](https://github.com/python/typeshed/issues/1339#issuecomment-974165907).

**Reply by whitphx** [reply]
> It would also be useful for typeshed to be able to test complex signatures. 

I totally agree, and I want it so much.
It's being done in the TypeScript's DefinitelyTyped repo like [this](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/esrever/esrever-tests.ts). There should be the like in Python.

**Reply by Viicos** [reply]
> The type checker-specific error codes are probably overkill though, so let's drop that.

My current workaround is to use `assert_type` knowing the type is wrong, and add a `type: ignore` comment. With `mypy` at least, you can explicitly error on unused `type: ignore` comments so that isn't accidentally used

**Comment by graingert** [comment, marked as answer]
For those coming to this discussion mypy now has native support for `assert_type` https://mypy.readthedocs.io/en/stable/error_code_list.html#check-types-in-assert-type-assert-type

```python
from typing_extensions import assert_type

assert_type([1], list[int])  # OK

assert_type([1], list[str])  # Error
```

**Reply by whitphx** [reply]
Thank you!

Also it's available from the standard lib `type` since 3.11 😄
https://docs.python.org/3/library/typing.html#typing.assert_type

```python
from typing import assert_type
```

**Reply by graingert** [reply]
@whitphx can you select this as an answer?

**Reply by whitphx** [reply]
Oops sorry I forgot it.
Done 👍

**Reply by RBerga06** [reply]
This works for ordinary tests, but there's no way of doing something similar for e.g. `@pytest.mark.parametrize(...)`d tests, right?

(note: I'm not using mypy, but instead pyright and ideally I'd like to ensure compatibility with ty as well...)
